### PR TITLE
Task/kab163/setup gitlab ci

### PR DESCRIPTION
**Pull request by kab163**
**Note:** Running tests on Quartz for now until a Radiuss bank is created on Ruby (then jobs will be switched back). 

**Question:** Should I include an update in the RAJAPerf README file that talks about Gitlab CI stuff?

Failing Gitlab jobs:
- [x] Error with ICPC job on Quartz - "Access violation or stack overflow. Please contact Intel Support for assistance."
         Fix: Removed icpc@18 job, may add updated icpc version job later
                Changed any intel jobs to run with "make -j 16" instead of default for LC machine.
- [x] Most Lassen job errors seem to revolve around HALOEXCHANGE_FUSED (memory leak errors + seg faulting errors)
         Update: Seg faulting and memory leak errors seemed to go away at one point, but have since come back - will need to talk to someone about what's happening
          Fix: The issues seem to be the result of differences between the clang and clang-ibm compilers. To solve this problem for now, we got rid of the clang-ibm jobs and replaced with the regular clang compiler.
- [x] One Lassen error on XL+gcc cuda job - "CUDAassert: CUDA driver version is insufficient for CUDA runtime version"
         Fix: Changed to cuda v. 11, skipping the actual run of the binary since not necessary
                Changed blueos + cuda jobs to have added flag during run

- [x] Make sure that blueos+cuda added flag is actually working correctly and not just being ignored.
          Fix: Added it back in the previous location I had it, but added "lrun" before the flag so it was correctly used.

### CODE REVIEW

**Comment by kab163** [comment]
LGTM

**Comment by kab163** [comment]
LGTM

**Comment by kab163** [comment]
LGTM

**Comment by kab163** [comment]
LGTM

**Comment by kab163** [comment]
LGTM

**Comment by kab163** [comment]
LGTM

**Comment by kab163** [comment]
LGTM

**Comment by kab163** [comment]
LGTM

**Comment by kab163** [comment]
LGTM

**Comment by kab163** [comment]
LGTM

**Review by adrienbernede — COMMENTED** [review]
Several small comments.

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+# In practice q_*, l_* and b_* stages are independently run and start immediately.
+
+stages:
+  - q_allocate_resources
+  - q_build_and_test
+  - q_release_resources
+  - l_build_and_test
+  - b_build_and_test
+  - c_build_and_test
+  - multi_project
```
[the remark this answers is not in the document]
Comment (kab163) [reply .gitlab-ci.yml:53]: no, i'm pretty sure it's not. I can remove it!

**Review by adrienbernede — APPROVED** [review]

**Comment by rhornung67** [comment]
@kab163 where is the '-finline-limit=20000' argument coming from that is being passed to the clang-11 compiler on lassen. I don't think that should be there.

**Comment by kab163** [comment]
> @kab163 where is the '-finline-limit=20000' argument coming from that is being passed to the clang-11 compiler on lassen. I don't think that should be there. Specifically for the ibm-clang11-gcc-8-cuda build. The warnings don't appear for other clang 11 builds.

It was being added for cuda jobs that don't use the XL compiler (left over from the other build_and_run.sh job script I used to create this current one). I changed it so that cuda jobs that are not using XL and that are not using clang will still have the flag, though perhaps I should just get rid of it all together.

**Comment by kab163** [comment]
@rhornung67 - This is ready for review!

**Review by rhornung67 — APPROVED** [review]
Looks good. Thanks @kab163

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 2]
```diff
@@ -0,0 +1,10 @@
+# Summary (Write a short headline summary of PR)
+
+- This PR is a (refactoring, bugfix, feature, something else)
+- It does the following (modify list as needed):
+  - Modifies/refactors (class or method) (how?)
+  - Fixes (issue number(s))
+  - Adds (specific feature) at the request of (project or person)
+
+*IMPORTANT NOTE! Remember to comment "LGTM" after pushing a commit to trigger the Gitlab CI.
```
[the remark this answers is not in the document]
Comment (kab163) [reply .github/PULL_REQUEST_TEMPLATE.md]: Good catch. Removed that bit.

**Review by rhornung67 — APPROVED** [review]

File: README.md [hunk 3]
```diff
@@ -711,6 +712,14 @@ above.
 
 * * *
 
+# Continuous Integration
+
+RAJAPerf Suite uses continuous integration to ensure that changes added to the repository are well integrated and tested for compatability with the rest of the existing code base. Our CI tests incude a variety of vetted configurations that run on different LC machines.
+
+RAJAPerf Suite shares its Gitlab CI workflow with other projects. The documentation is therefore [shared](https://radiuss-ci.readthedocs.io/en/latest/uberenv.html#ci).
```
Comment (adrienbernede) [inline README.md:719]: All the radiuss-ci documentation is about this workflow. So pointing to a more general reference.

```suggestion
RAJAPerf Suite shares its Gitlab CI workflow with other projects. The documentation is therefore [shared](https://radiuss-ci.readthedocs.io/en/latest).
```

**Review by adrienbernede — APPROVED** [review]
I know for a fact that this can be quite an adventure to get this working for all the specs.
This is not "just" about implementing the CI, but also getting the build and test to pass.
Great contribution @kab163!

File: .gitlab/ruby-templates.yml [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+# This is the shared configuration of jobs for ruby
+
+####
+# In pre-build phase, allocate a node for builds
+.on_ruby:
+  tags:
+    - shell
+    - ruby
+  rules:
+    - if: '$CI_COMMIT_BRANCH =~ /_qnone/ || $ON_RUBY == "OFF"' #run except if ...
```
[the remark this answers is not in the document]
Comment (kab163) [reply .gitlab/ruby-templates.yml:18]: Thanks!

**Review by rhornung67 — APPROVED** [review]

**Pull request merged**